File: src/hash/lc438.go
package hash

func findAnagrams(s string, p string) []int {
	if len(s) < len(p) {
		return []int{}
	}

	wndChr2cnt := make(map[byte]int)
	pChr2cnt := make(map[byte]int)

	for i := range len(p) {
		wndChr2cnt[s[i]]++
		pChr2cnt[p[i]]++
	}

	equal := func() bool {
		for chr, cnt := range pChr2cnt {
			if wndCnt, ok := wndChr2cnt[chr]; !ok || wndCnt != cnt {
				return false
			}
		}
		return true
	}

	ans := make([]int, 0)
	for r := len(p); r < len(s); r++ {
		l := r - len(p)
		if equal() {
			ans = append(ans, l)
		}
		wndChr2cnt[s[l]]--
		wndChr2cnt[s[r]]++
	}

	if equal() {
		ans = append(ans, len(s)-len(p))
	}

	return ans
}

var FindAnagrams = findAnagrams

File: src/hash/lc438.ts
function findAnagrams(s: string, p: string): number[] {
  if (s.length < p.length) {
    return [];
  }

  const wndChr2cnt: Record<string, number> = {};
  const pChr2cnt: Record<string, number> = {};

  for (let i = 0; i < p.length; i++) {
    const sChar = s[i];
    const pChar = p[i];
    wndChr2cnt[sChar] = (wndChr2cnt[sChar] || 0) + 1;
    pChr2cnt[pChar] = (pChr2cnt[pChar] || 0) + 1;
  }

  const equal = (): boolean => {
    for (const chr in pChr2cnt) {
      if (pChr2cnt[chr] !== (wndChr2cnt[chr] || 0)) {
        return false;
      }
    }
    return true;
  };

  const ans: number[] = [];
  for (let r = p.length; r < s.length; r++) {
    const l = r - p.length;
    if (equal()) {
      ans.push(l);
    }
    const leftChar = s[l];
    const rightChar = s[r];
    wndChr2cnt[leftChar]--;
    wndChr2cnt[rightChar] = (wndChr2cnt[rightChar] || 0) + 1;
  }

  if (equal()) {
    ans.push(s.length - p.length);
  }

  return ans;
}

export default findAnagrams;
